pixy/groupreg.go

```diff
@@ -0,0 +1,218 @@
+// `Topics()` channel. It also watches for other group members that join/leave
+// the group. Updated group member subscriptions are sent down to the
+// `MembershipChanges()` channel.
+type consumerGroupRegistry struct {
+	baseCID             *sarama.ContextID
+	group               string
+	config              *Config
+	groupZNode          *kazoo.Consumergroup
+	groupMemberZNode    *kazoo.ConsumergroupInstance
+	topicsCh            chan []string
+	membershipChangesCh chan map[string][]string
+	stoppingCh          chan none
+	wg                  sync.WaitGroup
+}
+
+func spawnConsumerGroupRegister(group, memberID string, config *Config, kazooConn *kazoo.Kazoo) *consumerGroupRegistry {
+	groupZNode := kazooConn.Consumergroup(group)
+	groupMemberZNode := groupZNode.Instance(memberID)
+	cgr := &consumerGroupRegistry{
+		baseCID:             sarama.RootCID.NewChild(fmt.Sprintf("registry:%s", group)),
+		group:               group,
+		config:              config,
+		groupZNode:          groupZNode,
+		groupMemberZNode:    groupMemberZNode,
+		topicsCh:            make(chan []string),
+		membershipChangesCh: make(chan map[string][]string),
+		stoppingCh:          make(chan none),
+	}
+	spawn(&cgr.wg, cgr.watcher)
+	spawn(&cgr.wg, cgr.register)
+	return cgr
+}
+
+func (cgr *consumerGroupRegistry) topics() chan<- []string {
+	return cgr.topicsCh
+}
+
+func (cgr *consumerGroupRegistry) membershipChanges() <-chan map[string][]string {
+	return cgr.membershipChangesCh
+}
+
+func (cgr *consumerGroupRegistry) claimPartition(cid *sarama.ContextID, topic string, partition int32, cancelCh <-chan none) func() {
+	if !retry(func() error { return cgr.groupMemberZNode.ClaimPartition(topic, partition) }, nil,
+		fmt.Sprintf("<%s> failed to claim partition", cid), cgr.config.Consumer.BackOffTimeout, cancelCh,
+	) {
+		log.Infof("<%s> partition claimed", cid)
+	}
+	return func() {
+		if !retry(func() error { return cgr.groupMemberZNode.ReleasePartition(topic, partition) },
+			func(err error) bool { return err != nil && err != kazoo.ErrPartitionNotClaimed },
+			fmt.Sprintf("<%s> failed to release partition", cid), cgr.config.Consumer.BackOffTimeout, cancelCh,
+		) {
+			log.Infof("<%s> partition released", cid)
+		}
+	}
+}
+
+func (cgr *consumerGroupRegistry) stop() {
+	close(cgr.stoppingCh)
+	cgr.wg.Wait()
+}
+
+// partitionOwner returns the id of the consumer group member that has claimed
+// the specified topic/partition.
+func (cgr *consumerGroupRegistry) partitionOwner(topic string, partition int32) (string, error) {
+	owner, err := cgr.groupZNode.PartitionOwner(topic, partition)
+	if err != nil {
+		return "", err
+	}
+	if owner == nil {
+		return "", nil
+	}
+	return owner.ID, nil
+}
+
+// watcher keeps an eye on the consumer group membership changes, that is when
+// members join/leave the group and sends notifications about such changes down
+// to the `membershipChangesCh` channel.
+func (cgr *consumerGroupRegistry) watcher() {
```

Comment:
It looks like this function will always send one change notification when it is first called, even if there haven't been changes. Is that intentional? If so can you add that to the function comment?

The reason it looks like that to me is that the function will block on the first select statement only until `cgr.config.Consumer.RebalanceDelay` time elapses, and then this function will call `cgr.sendMembershipUpdate`. My understanding of `WatchInstances` is that it will not block, either. So it looks like this function will send one change notification before blocking in the second select statement to wait for the first change. Is my understanding correct?